File: backend/services/student_service.py
from io import BytesIO
import pandas as pd
from extensions import db
from repositories.user_repository import UserRepository
from repositories.class_repository import ClassRepository
from repositories.student_repository import StudentRepository
from errors import NotFoundError, ConflictError

user_repo    = UserRepository()
class_repo   = ClassRepository()
student_repo = StudentRepository()


def search_students(q: str, class_id: int = None) -> list:
    students = user_repo.search_students(q)
    for s in students:
        s["already_in_class"] = (class_id is not None and s["class_id"] == class_id)
        if s.get("dob"):
            s["dob"] = str(s["dob"])
    return students


def add_to_class(class_id: int, teacher_id: int, username: str) -> str:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    student = user_repo.find_by_username_and_role(username, "student")
    if not student:
        raise NotFoundError(f'Không tìm thấy học sinh "{username}"')
    if student.class_id == class_id:
        raise ConflictError("Học sinh đã trong lớp này")
    student_repo.add_to_class(student, class_id)
    db.session.commit()
    return student.full_name


def upload_students(class_id: int, teacher_id: int, file_bytes: bytes) -> dict:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    df = pd.read_excel(BytesIO(file_bytes))
    df.columns = df.columns.str.strip().str.lower()
    if "username" not in df.columns:
        raise ValueError("File thiếu cột: username")

    success_count = 0
    errors        = []
    for _, row in df.iterrows():
        username = str(row["username"]).strip()
        if not username or username == "nan":
            errors.append("Bỏ qua dòng trống")
            continue
        student = user_repo.find_by_username_and_role(username, "student")
        if not student:
            errors.append(f'Không tìm thấy học sinh "{username}"')
            continue
        if student.class_id == class_id:
            errors.append(f'"{username}" đã trong lớp này')
            continue
        student_repo.add_to_class(student, class_id)
        success_count += 1

    if success_count == 0:
        raise NotFoundError("Không tìm thấy học sinh")
    db.session.commit()
    return {"count": success_count, "errors": errors}


def remove_from_class(class_id: int, teacher_id: int, student_id: int) -> None:
    if not class_repo.find_by_id_and_teacher(class_id, teacher_id):
        raise NotFoundError("Lớp không tồn tại")
    student = user_repo.find_by_id(student_id)
    if not student or student.class_id != class_id:
        raise NotFoundError("Học sinh không trong lớp này")
    student_repo.remove_from_class(student)
    db.session.commit()


def get_student_results(student_id: int) -> dict:
    student = student_repo.find_by_id(student_id)
    if not student:
        raise NotFoundError("Học sinh không tồn tại")
    results = student_repo.get_results(student_id)
    for r in results:
        if r.get("submittedAt"):
            r["submittedAt"] = str(r["submittedAt"])
        if r.get("score") is not None:
            r["score"] = float(r["score"])
    return {"studentName": student.full_name, "results": results}


def get_student_progress(student_id: int) -> list:
    rows = student_repo.get_progress(student_id)
    for r in rows:
        r["submittedAt"] = str(r["submittedAt"]) if r.get("submittedAt") else None
        r["score"]       = float(r["score"] or 0)
    return rows


def get_student_dashboard(student_id: int, show_all: bool, date_from=None, date_to=None) -> dict:
    teacher_id = student_repo.get_teacher_id_for_student(student_id)
    if not teacher_id:
        return {
            "exams": [], "scores": [], "ranking": [],
            "progress": {"totalExams": 0, "done": 0, "avg": None},
            "announcements": [],
        }

    df = None if show_all else date_from
    dt = None if show_all else date_to

    lesson_rows  = student_repo.get_lessons_with_stats(student_id, df, dt)
    ranking_rows = student_repo.get_class_ranking(teacher_id, df, dt)
    announcements = student_repo.get_announcements_for_student(student_id)
    for a in announcements:
        a["created_at"] = str(a["created_at"]) if a.get("created_at") else None

    exams  = []
    scores = []
    for r in lesson_rows:
        done  = bool(r["done"])
        score = None
        if done and r["total_questions"]:
            score = round(float(r["correct_questions"] or 0) / float(r["total_questions"]) * 10, 1)
        exams.append({
            "examId":     r["exam_id"],  "examName":   r["exam_name"],
            "lessonTitle": r["lesson_name"], "done": done, "score": score,
            "deadline":   str(r["deadline"])  if r.get("deadline")  else None,
            "openTime":   str(r["open_time"]) if r.get("open_time") else None,
        })
        if done and score is not None:
            scores.append({"examName": r["exam_name"], "score": score})

    done_count = sum(1 for e in exams if e["done"])
    avg        = round(sum(s["score"] for s in scores) / len(scores), 1) if scores else None
    ranking    = [
        {
            "studentId": r["student_id"], "name": r["student_name"],
            "avg": round(float(r["correct_questions"] or 0) / float(r["total_questions"]) * 10, 1)
                   if r["total_questions"] else 0,
        }
        for r in ranking_rows
    ]

    return {
        "exams": exams, "scores": scores, "ranking": ranking,
        "progress": {"totalExams": len(exams), "done": done_count, "avg": avg},
        "announcements": announcements,
    }

File: backend/src/services/student.service.js
import XLSX from 'xlsx';
import * as userRepo from '../repositories/user.repository.js';
import * as classRepo from '../repositories/class.repository.js';
import * as studentRepo from '../repositories/student.repository.js';
import { NotFoundError, ConflictError } from '../utils/error.utils.js';
import { formatDate, formatDateTime } from '../utils/date.utils.js';

export const searchStudents = async (q, classId = null) => {
  const students = await userRepo.searchStudents(q);
  return students.map(s => ({
    ...s,
    already_in_class: classId != null && s.class_id === classId,
    dob: formatDate(s.dob),
  }));
};

export const addToClass = async (classId, teacherId, username) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');
  const student = await userRepo.findByUsernameAndRole(username, 'student');
  if (!student) throw new NotFoundError(`Không tìm thấy học sinh "${username}"`);
  if (student.class_id === classId) throw new ConflictError('Học sinh đã trong lớp này');
  if (student.class_id != null) throw new ConflictError('Học sinh đã thuộc lớp khác');
  await studentRepo.addToClass(student.id, classId);
  return student.full_name;
};

export const uploadStudents = async (classId, teacherId, fileBuffer) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');

  const workbook = XLSX.read(fileBuffer, { type: 'buffer' });
  const sheet = workbook.Sheets[workbook.SheetNames[0]];
  const rawRows = XLSX.utils.sheet_to_json(sheet, { defval: '' });
  const rows = rawRows.map(r => {
    const n = {};
    for (const [k, v] of Object.entries(r)) n[k.trim().toLowerCase()] = v;
    return n;
  });

  if (!rows.length || !('username' in rows[0])) throw new Error('File thiếu cột: username');

  let successCount = 0;
  const errors = [];
  for (const row of rows) {
    const username = String(row.username || '').trim();
    if (!username || username === 'nan') { errors.push('Bỏ qua dòng trống'); continue; }
    const student = await userRepo.findByUsernameAndRole(username, 'student');
    if (!student) { errors.push(`Không tìm thấy học sinh "${username}"`); continue; }
    if (student.class_id === classId) { errors.push(`"${username}" đã trong lớp này`); continue; }
    await studentRepo.addToClass(student.id, classId);
    successCount++;
  }

  if (successCount === 0) throw new NotFoundError('Không tìm thấy học sinh');
  return { count: successCount, errors };
};

export const generateSampleExcel = () => {
  const rows = [
    { username: 'hocsinh01' },
    { username: 'hocsinh02' },
  ];
  const wb = XLSX.utils.book_new();
  const ws = XLSX.utils.json_to_sheet(rows);
  XLSX.utils.book_append_sheet(wb, ws, 'Sheet1');
  return XLSX.write(wb, { type: 'buffer', bookType: 'xlsx' });
};

export const removeFromClass = async (classId, teacherId, studentId) => {
  if (!await classRepo.findByIdAndTeacher(classId, teacherId)) throw new NotFoundError('Lớp không tồn tại');
  const student = await userRepo.findById(studentId);
  if (!student || student.class_id !== classId) throw new NotFoundError('Học sinh không trong lớp này');
  await studentRepo.removeFromClass(studentId);
};

export const getStudentResults = async (studentId) => {
  const student = await studentRepo.findById(studentId);
  if (!student) throw new NotFoundError('Học sinh không tồn tại');
  const results = await studentRepo.getResults(studentId);
  return {
    studentName: student.full_name,
    results: results.map(r => ({
      ...r,
      submittedAt: formatDateTime(r.submittedAt),
      score: r.score != null ? parseFloat(r.score) : null,
    })),
  };
};

export const getStudentProgress = async (studentId) => {
  const rows = await studentRepo.getProgress(studentId);
  return rows.map(r => ({
    ...r,
    submittedAt: formatDateTime(r.submittedAt),
    score: parseFloat(r.score || 0),
  }));
};

export const getStudentDashboard = async (studentId, showAll, dateFrom = null, dateTo = null) => {
  const teacherInfo = await studentRepo.getTeacherInfoForStudent(studentId);
  if (!teacherInfo) {
    return {
      exams: [], scores: [], ranking: [],
      progress: { totalExams: 0, done: 0, avg: 0 },
      announcements: [], teacher: null,
    };
  }
  const teacherId = teacherInfo.teacher_id;
  const df = showAll ? null : dateFrom;
  const dt = showAll ? null : dateTo;

  const lessonRows = await studentRepo.getLessonsWithStats(studentId, df, dt);
  const rankingRows = await studentRepo.getClassRanking(teacherId, df, dt);
  const announcements = await studentRepo.getAnnouncementsForStudent(studentId);
  const announcementsFormatted = announcements.map(a => ({
    ...a,
    created_at: formatDate(a.created_at),
  }));

  const exams = [];
  const scores = [];
  for (const r of lessonRows) {
    const done = Boolean(r.done);
    let score = null;
    if (done && r.total_questions) {
      score = parseFloat((parseFloat(r.correct_questions || 0) / parseFloat(r.total_questions) * 10).toFixed(1));
    }
    exams.push({
      examId: r.exam_id, examName: r.exam_name, lessonTitle: r.lesson_name,
      done, score,
      deadline: formatDateTime(r.deadline),
      openTime: formatDateTime(r.open_time),
    });
    if (done && score != null) {
      scores.push({ examName: r.exam_name, score, _created_at: r.exam_created_at });
    }
  }
  scores.sort((a, b) => String(a._created_at || '').localeCompare(String(b._created_at || '')));
  const scoresClean = scores.map(({ _created_at, ...s }) => s);

  const doneCount = exams.filter(e => e.done).length;
  const avg = scoresClean.length
    ? parseFloat((scoresClean.reduce((s, x) => s + x.score, 0) / scoresClean.length).toFixed(1))
    : null;

  const ranking = rankingRows.map(r => ({
    studentId: r.student_id,
    name: r.student_name,
    avg: r.total_questions
      ? parseFloat((parseFloat(r.correct_questions || 0) / parseFloat(r.total_questions) * 10).toFixed(1))
      : 0,
  }));

  return {
    exams, scores: scoresClean, ranking,
    progress: { totalExams: exams.length, done: doneCount, avg },
    announcements: announcementsFormatted,
    teacher: {
      fullName: teacherInfo.full_name,
      email: teacherInfo.email,
      phone: teacherInfo.phone,
    },
  };
};
